### PR TITLE
Fixed #31217 -- Made QuerySet.values()/values_list() group by not selected annotations with aggregations used in order_by().

### PR DESCRIPTION
**Pull request by felixxm**
ticket-31217

Regression in 59b4e99dd00b9c36d56055b889f96885995e4240.

Thanks Jon Dufresne for the report.

### CODE REVIEW

**Comment by felixxm** [comment]
@jdufresne Can you confirm that this patch works for you?

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
@carltongibson You need to check on PostgreSQL.

**Comment by carltongibson** [comment]
Ah, there's my failure. Ta!

**Comment by carltongibson** [comment]
Seems to make sense. 

The original query, prior to the regression: 

```
SELECT "aggregation_book"."id" 
...
GROUP BY "aggregation_book"."id", T4."age" 
```

Now: 

```
SELECT "aggregation_book"."id", COALESCE(MIN("aggregation_author"."age"), T4."age") AS "min_related_age" 
...
GROUP BY "aggregation_book"."id", T4."age" 
```

Does the extra `"min_related_age"` SELECT matter? (The `values_list()` call means it's not returned, and without that both the annotations would be selected anyways so...)

**Comment by felixxm** [comment]
@carltongibson Great catch! I added `copy()` to avoid mutating the list of selected annotations.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by jdufresne** [comment]
Thanks for writing the fix! My project still fails with this patch in place.

It looks like it doesn't handle the ordering by a `DESC` annotation.

```py
    def test_aggregation_order_by_desc_not_selected_annotation_values(self):
        books_qs = Book.objects.annotate(
            min_age=Min('authors__age'),
        ).annotate(
            min_related_age=Coalesce('min_age', 'contact__age'),
        ).order_by('-min_related_age').values_list('pk', flat=True)
        self.assertEqual(
            list(books_qs),
            [self.b6.pk, self.b5.pk, self.b2.pk, self.b1.pk, self.b3.pk, self.b4.pk],
        )
```

```
======================================================================
ERROR: test_aggregation_order_by_desc_not_selected_annotation_values (aggregation.tests.AggregateTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jon/devel/django/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.GroupingError: column "t4.age" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: ...RDER BY COALESCE(MIN("aggregation_author"."age"), T4."age") ...
                                                             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jon/devel/django/tests/aggregation/tests.py", line 1210, in test_aggregation_order_by_desc_not_selected_annotation_values
    list(books_qs),
  File "/home/jon/devel/django/django/db/models/query.py", line 258, in __len__
    self._fetch_all()
  File "/home/jon/devel/django/django/db/models/query.py", line 1261, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/home/jon/devel/django/django/db/models/query.py", line 184, in __iter__
    for row in compiler.results_iter(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size):
  File "/home/jon/devel/django/django/db/models/sql/compiler.py", line 1096, in results_iter
    results = self.execute_sql(MULTI, chunked_fetch=chunked_fetch, chunk_size=chunk_size)
  File "/home/jon/devel/django/django/db/models/sql/compiler.py", line 1144, in execute_sql
    cursor.execute(sql, params)
  File "/home/jon/devel/django/django/db/backends/utils.py", line 68, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/home/jon/devel/django/django/db/backends/utils.py", line 77, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/jon/devel/django/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
  File "/home/jon/devel/django/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/jon/devel/django/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: column "t4.age" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: ...RDER BY COALESCE(MIN("aggregation_author"."age"), T4."age") ...
                                                             ^
```

**Comment by felixxm** [comment]
@jdufresne Thanks for checking :+1:  I added regression tests for all cases (also with the `OrderBy()` expression). I just need to fix them ...

File: django/db/models/sql/query.py [hunk 1]
```diff
@@ -1928,8 +1928,17 @@ def set_group_by(self, allow_aliases=True):
         will be made automatically.
         """
         group_by = list(self.select)
-        if self.annotation_select:
-            for alias, annotation in self.annotation_select.items():
+        annotations_group_by = self.annotation_select.copy()
+        # Add not selected annotations used only in the ORDER BY clause.
+        if self.order_by and self.annotations:
+            for item in self.order_by:
+                if not hasattr(item, 'resolve_expression'):
```
Comment (felixxm) [inline django/db/models/sql/query.py]: We allow for all kind of expressions in `order_by()`. I wonder if there an easy way to check annotations used in expressions :thinking: @charettes Any ideas?

**Comment by felixxm** [comment]
I prepared an alternative approach that works in all cases :relaxed: . I believe that the previous attempt was in the wrong place because this issue is more niche than I thought, it only affects not select annotations which contains aggregations and are used in `order_by()` with `QuerySet.values()/values_list()` :male_detective:

Compiler omitted expressions with aggregations used in the `ORDER BY` clause to prevent:
```
psycopg2.errors.GroupingError: aggregate functions are not allowed in GROUP BY
```
which was fine before 59b4e99dd00b9c36d56055b889f96885995e4240 because we included `get_group_by_cols()` for all annotations. After 59b4e99dd00b9c36d56055b889f96885995e4240 such columns were missing. I used `expr.get_group_by_cols()` instead of `expr.get_source_expressions()` to avoid `psycopg2.errors.GroupingError` for expressions which contain aggregations.

@jdufresne Can you check again?

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
Tests failure on Oracle is not related with this change, all tests passed:
```
Ran 13468 tests in 4803.211s

OK (skipped=802, expected failures=6)
```
it's just a random failure on cleanup.

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -123,8 +123,8 @@ def get_group_by(self, select, order_by):
         for expr, (sql, params, is_ref) in order_by:
             # Skip References to the select clause, as all expressions in the
             # select clause are already part of the group by.
-            if not expr.contains_aggregate and not is_ref:
-                expressions.extend(expr.get_source_expressions())
+            if not is_ref:
+                expressions.extend(expr.get_group_by_cols())
```
Comment (charettes) [inline django/db/models/sql/compiler.py]: That makes sense, using `get_source_expressions` here was only performing a `contains_aggregate` check for the other level expression while `get_group_by_cols` does it recursively as the expression tree is walked.

https://github.com/django/django/blob/c1b24718e05ea474955777d7bc4d9d5634560cd5/django/db/models/expressions.py#L346-L352

Comment (charettes) [reply django/db/models/sql/compiler.py]: @felixxm makes me wonder if we want to do `source.get_group_by_cols(alias)` in `BaseExpression.get_group_by_cols`. The current implementation makes me think we might be suffering from an issue similar to [#31094](https://github.com/django/django/commit/5a4d7285bd10bd40d9f7e574a7c421eb21094858) if `Coalesce(Subquery)` is used instead. I'll submit a PR if it's the case.

**Comment by jdufresne** [comment]
This solves bug in my project :tada:

Thanks!

File: django/db/models/sql/compiler.py (hunk 2 above)
Comment (felixxm) [reply django/db/models/sql/compiler.py]: @charettes Thanks for checking :+1: Agreed, it looks suspicious, however I wasn't able to reproduce a similar issue (at least for now). I can play with this a bit more next week.

**Comment by felixxm** [comment]
@jdufresne Thanks for checking!

**Pull request merged**